doc/designs/application-specific-passwords.md

```diff
@@ -0,0 +1,283 @@
+from scratch, even a limitation such as reducing the effective permissions of
+the current user session (possibly according to the application policy) using a
+389-DS plugin probably will not be needed.
+
+#### HBAC for application-specific passwords
+
+Two-component principals for subaccounts, as described above, will be used for
+assigning HBAC rules to users logged in using application-specific passwords.
+These HBAC rules will be assigned by administrators just as the other ones.
+
+TBD: When an application-specific password is deleted, delete the associated
+HBAC rules automatically, or keep them?
+
+Optionally, a possibility for administrators to specify that users can log into
+a particular service only using the passwords from a specified LDAP subtree (or
+the primary password) can be provided.
+
+hbacrule plugin therefore will need to be extended with the ability to assign
+allowed LDAP subtrees to application-specific passwords as a part of HBAC rules.
+hbactest plugin will be extended to test this, and it will be used to test HBAC
+with application-specific passwords in general.
+
+### SSSD support for the aforementioned functionality
+
+SSSD will also be extended to support the aforementioned functionality. This
+will be done by a closely related, but separate extension of SSSD. On the
+side of FreeIPA, after SSSD is extended, interactions with SSSD using the new
+functionality must be tested in `ipatests/test_integration/test_sssd.py`.
+
+## Implementation
+
+The extension must be backward compatible, that means not breaking any existing
+functionality and passing all existing tests.
+
+As for changes to the data storage:
+* As described above, a new LDAP subtree will be created for
+  application-specific passwords, with objects holding them being grouped by
+  application, and their attributes being, besides an application-specific
+  password itself, a two-component Kerberos principal containing the application
+  name, and a backlink to the owning user.
+* New ACIs will be automatically created and maintained so that users will have
+  read-write access to objects holding their own application-specific passwords.
+* HBAC rules for subaccounts will be assigned in the same way as the other ones,
+  using the two-component principal.
+* Permitted LDAP subtrees for application-specific passwords will be stored as a
+  part of HBAC rules.
+
+The new additions to the data storage (including the new object and its
+attributes) must be specified in LDIF files in `install/share/`, probably in
+`60basev3.ldif`, or alternatively, in a new file.
+
+Analogically, such records probably will be added to a new file in 
+`install/updates`.
+
+Backup and restore is out of scope of this extension, but it will probably work
+with the new functionality.
+
+**New dependencies**: TBD, not expected for now
+
+## Feature management
+
+### UI
+
+It is not necessary that the new functionality be available in WebUI, but it
+would be good to do it.
+
```

Comment:
Although this is technically true, a UI will be the only way for most deployments to be really usable.
Some more consideration should be given on what should be the rough outlines of a UI.

Reply:
All right, I will make it available in the WebUI as well. I will set it out when the outstanding questions will be resolved and I will dive into the detailed phase of the design.